### PR TITLE
add gpt-4-turbo

### LINKED ISSUE #347: GPT-4-Turbo
**Is your feature request related to a problem? Please describe.**
The current gpt 4 implementation is expensive

**Describe the solution you'd like**
Use the new gpt-4-1106-preview which is approximately 3x cheaper

### PR DESCRIPTION
**Pull request by cpacker**
Closes #347

### CODE REVIEW

File: memgpt/config.py [hunk 1]
```diff
@@ -28,6 +28,10 @@
 
 model_choices = [
     questionary.Choice("gpt-4"),
+    questionary.Choice(
+        "gpt-4-turbo (developer preview)",
+        value="gpt-4-1106-preview",
+    ),
     questionary.Choice(
         "gpt-3.5-turbo (experimental! function-calling performance is not quite at the level of gpt-4 yet)",
```
Comment (cpacker) [inline memgpt/config.py:36]: Should we make the choice not be `gpt-3.5-turbo`, but instead `gpt-3.5-turbo-16k`? I think it's generally available at this point, and non-16k is actually only 4k context limit which is really small for MemGPT.

File: memgpt/cli/cli_config.py [hunk 2]
```diff
@@ -95,9 +95,9 @@ def configure():
     if use_openai or use_azure:
         model_options = []
         if use_openai:
-            model_options += ["gpt-3.5-turbo", "gpt-3.5", "gpt-4"]
+            model_options += ["gpt-3.5-turbo", "gpt-4", "gpt-4-1106-preview"]
         default_model = questionary.select(
-            "Select default model (recommended: gpt-4):", choices=["gpt-3.5-turbo", "gpt-3.5", "gpt-4"], default="gpt-4"
+            "Select default model (recommended: gpt-4):", choices=["gpt-3.5-turbo", "gpt-4", "gpt-4-1106-preview"], default="gpt-4"
```
Comment (cpacker) [inline memgpt/cli/cli_config.py:100]: I made the diff here too because I just grep'd for `gpt-3.5-turbo`, but I'm not sure if this is needed too.

**Comment by cpacker** [comment]
Also, we really should change `LLM_MAX_TOKENS` dynamically in `constants.py` to made the model selected. At least when using OpenAI, for local LLMs we can expect people to change this manually for now (but also we should leave it at 8k for this reason because most local LLM people are using Mistral 7B finetunes (8k).

File: memgpt/config.py (hunk 1 above)
Comment (vivi) [reply memgpt/config.py:36]: That sounds good to me.

**Review by vivi — APPROVED** [review]
lgtm pending gpt-3.5-16k

**Comment by sarahwooders** [comment]
I get an error: 
```
╰─ memgpt run                                                                                                                               ─╯
? Would you like to select an existing agent? No
Creating new agent...
Created new agent agent_9.
Hit enter to begin (will request first MemGPT message)

step() failed
user_message = None
error = HTTP code 403 from API ()
HTTP code 403 from API ()
An exception ocurred when running agent.step():
Traceback (most recent call last):
  File "/Users/sarahwooders/repos/MemGPT/env/lib/python3.10/site-packages/openai/api_requestor.py", line 765, in _interpret_response_line
    data = json.loads(rbody)
  File "/usr/local/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/local/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/Cellar/python@3.10/3.10.13/Frameworks/Python.framework/Versions/3.10/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/main.py", line 613, in run_agent_loop
    new_messages, user_message, skip_next_user_input = await process_agent_step(user_message, no_verify)
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/main.py", line 589, in process_agent_step
    new_messages, heartbeat_request, function_failed, token_warning = await memgpt_agent.step(
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/agent.py", line 1136, in step
    raise e
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/agent.py", line 1072, in step
    response = await get_ai_reply_async(model=self.model, message_sequence=input_message_sequence, functions=self.functions)
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/agent.py", line 164, in get_ai_reply_async
    raise e
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/agent.py", line 145, in get_ai_reply_async
    response = await acreate(
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/openai_tools.py", line 115, in wrapper
    raise e
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/openai_tools.py", line 95, in wrapper
    return await func(*args, **kwargs)
  File "/Users/sarahwooders/repos/memgpt-main/MemGPT/memgpt/openai_tools.py", line 135, in acompletions_with_backoff
    return await openai.ChatCompletion.acreate(**kwargs)
  File "/Users/sarahwooders/repos/MemGPT/env/lib/python3.10/site-packages/openai/api_resources/chat_completion.py", line 45, in acreate
    return await super().acreate(*args, **kwargs)
  File "/Users/sarahwooders/repos/MemGPT/env/lib/python3.10/site-packages/openai/api_resources/abstract/engine_api_resource.py", line 219, in acreate
    response, _, api_key = await requestor.arequest(
  File "/Users/sarahwooders/repos/MemGPT/env/lib/python3.10/site-packages/openai/api_requestor.py", line 384, in arequest
    resp, got_stream = await self._interpret_async_response(result, stream)
  File "/Users/sarahwooders/repos/MemGPT/env/lib/python3.10/site-packages/openai/api_requestor.py", line 738, in _interpret_async_response
    self._interpret_response_line(
  File "/Users/sarahwooders/repos/MemGPT/env/lib/python3.10/site-packages/openai/api_requestor.py", line 767, in _interpret_response_line
    raise error.APIError(
openai.error.APIError: HTTP code 403 from API ()
? Retry agent.step()? (Y/n)
```

**Comment by cpacker** [comment]
@sarahwooders seeing the JSON decode error, is there a chance that your env vars for local llm are still set?

Works for me atm:

image
image

**Comment by sarahwooders** [comment]
Yeah my mistake. Lgtm!

**Pull request merged**